experimental/trusted_shuffler/trusted_shuffler/src/lib.rs

```diff
@@ -16,4 +16,131 @@
+    O: Send + Sync + Future<Output = Response>,
+{
+    pub fn new(anonymity_value: usize, request_sender: F) -> Self {
+        Self {
+            anonymity_value,
+            current_requests: Arc::new(Mutex::new(vec![])),
+            request_sender,
+        }
+    }
+
+    // Asynchronously handles an incoming request.
+    pub async fn invoke(&self, request: Request) -> anyhow::Result<Response> {
+        let (sender, receiver) = oneshot::channel();
+
+        let request = Message {
+            data: request,
+            response_sender: sender,
+        };
+
+        // Check if the request batch is filled and create a separate task for shuffling requests.
+        {
+            let mut current_requests = self
+                .current_requests
+                .lock()
+                .map_err(|error| anyhow!("Couldn't lock current messages mutex: {:?}", error))?;
+            current_requests.push(request);
+
+            if current_requests.len() >= self.anonymity_value {
+                // Replace current requests with an empty vector.
+                let requests = replace(current_requests.deref_mut(), vec![]);
+
+                // Shuffle current requests and send them to the backend.
+                let request_sender = self.request_sender.clone();
+                tokio::spawn(async move {
+                    if let Err(error) =
+                        TrustedShuffler::shuffle_requests(requests, request_sender).await
+                    {
+                        log::error!("Shuffling error: {:?}", error);
+                    }
+                });
+            }
+        }
+
+        receiver.await.context("Couldn't receive response")
+    }
+
+    // Shuffles requests and sends them to the backend using the
+    // [`TrustedShuffler::request_sender`].
+    async fn shuffle_requests(mut requests: Vec<Message>, request_sender: F) -> anyhow::Result<()> {
+        requests.shuffle(&mut thread_rng());
+
+        let response_futures: Vec<_> = requests
+            .drain(..)
+            .into_iter()
+            .map(|request| {
+                let request_sender_clone = request_sender.clone();
+                async move {
+                    let response = (request_sender_clone)(request.data).await;
+                    Message {
+                        data: response,
+                        response_sender: request.response_sender,
+                    }
+                }
+            })
+            .collect();
+        let responses = join_all(response_futures).await;
+
+        Self::shuffle_responses(responses)
+    }
+
+    // Shuffles received backend responses and sends them back to the client async tasks using
+    // corresponding channels defined in the [`Message::response_sender`].
+    fn shuffle_responses(mut responses: Vec<Message>) -> anyhow::Result<()> {
+        responses.shuffle(&mut thread_rng());
+
+        for response in responses.drain(..) {
```

Comment:
Does it not work just removing drain then?

```suggestion
        for response in responses {
```

Reply:
Done